File: src/client/controller/subpages/LobbyController.java
package client.controller.subpages;

import client.controller.ClientApplicationController;
import client.model.BoggleApp.LobbyUser;
import client.model.subpages.GameRoomModel;
import client.model.subpages.LobbyModel;
import client.view.ClientApplicationView;
import client.view.subpages.LobbyView;
import org.omg.CORBA.BooleanHolder;

import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;

public class LobbyController {

    private LobbyModel model;
    private LobbyView view;

    private ClientApplicationController parent;

    Thread timerThread;

    private boolean exitLobby;

    public LobbyController(LobbyModel lobbyModel, LobbyView lobbyView, ClientApplicationController parent) {
        this.model = lobbyModel;
        this.view = lobbyView;
        this.parent = parent;

        timerThread = new Thread(new Runnable() {
            @Override
            public void run() {
                timer();
            }
        });
        timerThread.start();

        view.setExitLobbyListener(new ExitListener());
    }

    public void timer() {
        view.setLblTimerTxt("00:" + 10000 / 1000);
        long timerVal = -1;
        LobbyUser[] usersInLobby;
        LobbyUser[] tempUsersInLobby;
        exitLobby = false;
        try {
            model.getWfImpl().attemptJoin(model.getUsername());
            usersInLobby = model.getUsersInLobby();
            populateLobby(usersInLobby);

            BooleanHolder startLobby = new BooleanHolder(false);
            while (timerVal != 0) {
                if (exitLobby)
                    break;
                timerVal = model.getWfImpl().getCurrLobbyTimerValue(startLobby);
                String formattedTimer = String.format("%02d:%02d:%02d",
                        (timerVal / 3600000) % 60,
                        (timerVal / 60000) % 60,
                        (timerVal / 1000) % 60);
                System.out.println("TIMER VALUE: " + formattedTimer);

                tempUsersInLobby = model.getUsersInLobby();
                if (tempUsersInLobby.length != usersInLobby.length) {
                    System.out.println("reached");
                    usersInLobby = tempUsersInLobby;
                    populateLobby(usersInLobby);
                }

                view.setLblTimerTxt(formattedTimer);

            }
            if (startLobby.value) {
                System.out.println("start lobby is valid");
                parent.getView().showGameRoom();
                parent.stopMusic();
                parent.playGameMusic();
                new GameRoomController(new GameRoomModel(model.getUsername(), model.getWfImpl()),
                        parent.getView().getGameRoomView(), parent);
            } else {
                parent.getView().showHome();
                parent.playDefaultMusic();
                parent.getView().showButtons();
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void populateLobby(LobbyUser[] users) {
        view.removePlayersInUserPanel();
        for (LobbyUser user : users) {
            view.addPlayerInUserPanel(user.pfpAddress, user.username);
        }
    }

    public class ExitListener implements ActionListener {
        @Override
        public void actionPerformed(ActionEvent e) {
            timerThread.interrupt();
            exitLobby = true;
            model.getWfImpl().exitLobby(model.getUsername());
            parent.getView().showHome();
            parent.getView().setNavLocationText("Home");
            parent.getView().showButtons();
            parent.stopMusic();
            parent.playDefaultMusic();

        }
    }
}

File: src/Client_Python/yuen_client/Controller/subpages/LobbyController.py
'''import threading

class LobbyController:
    def __init__(self, lobby_model, lobby_view, parent):
        self.model = lobby_model
        self.view = lobby_view
        self.parent = parent

        #Thread nT = new Thread(new Runnable() {
        #    @Override
        #    public void run() {
        #        timer();
        #    }
        #});
        #nT.start();

        nT = threading.Thread(target=self.timer)
        nT.start()

    def timer(self):
        self.view.set_lbl_timer_txt("00:" + str(10000 // 1000))
        timer_val = -1
        users_in_lobby = None
        temp_users_in_lobby = None
        try:
            self.model.get_wf_impl().attempt_join(self.model.get_username())
            users_in_lobby = self.model.get_users_in_lobby()
            self.populate_lobby(users_in_lobby)

            start_lobby = False
            while timer_val != 0:
                timer_val, start_lobby = self.model.get_wf_impl().get_curr_lobby_timer_value()
                print("TIMER VALUE:", timer_val)

                users_in_lobby = self.model.get_users_in_lobby()
                for user in users_in_lobby:
                    print("user sent:", user.username, "|", user.pfp_address)

                temp_users_in_lobby = self.model.get_users_in_lobby()
                if len(temp_users_in_lobby) != len(users_in_lobby):
                    print("reached")
                    users_in_lobby = temp_users_in_lobby
                    self.populate_lobby(users_in_lobby)

                self.view.set_lbl_timer_txt("00:0" + str(timer_val // 1000))

            if start_lobby:
                self.parent.get_view().show_game_room()
                self.parent.stop_music()
                self.parent.play_game_music()
                GameRoomController(GameRoomModel(self.model.get_username(), self.model.get_wf_impl()),
                                   self.parent.get_view().get_game_room_view(), self.parent)
            else:
                self.parent.get_view().show_home()
                self.parent.play_default_music()
        except Exception as e:
            print(e)

    def populate_lobby(self, users):
        self.view.remove_players_in_user_panel()
        for user in users:
            self.view.add_player_in_user_panel(user.pfp_address, user.username)
'''